memtable/write_buffer_manager.cc

```diff
@@ -141,8 +188,43 @@ void WriteBufferManager::FreeMemWithCache(size_t mem) {
     cache_rep_->cache_allocated_size_ -= kSizeDummyEntry;
     dummy_size_.fetch_sub(kSizeDummyEntry, std::memory_order_relaxed);
   }
+
+  // Memory usage is under the limit after freeing the memory, signal the DBs if
+  // blocked.
+  if (!ShouldStall() && !queue_.empty()) {
+    EndWriteStall();
+  }
 #else
   (void)mem;
 #endif  // ROCKSDB_LITE
 }
+
+void WriteBufferManager::BeginWriteStall(DB* db) {
+  if (db && ShouldStall()) {
+    std::unique_lock<std::mutex> lock(mu_);
+    queue_.push_back(db);
+  }
+}
+
+// Called when memory is freed in FreeMem.
+void WriteBufferManager::EndWriteStall() {
+  // Get the db instances from the list and call WBMStallInterface::SignalDB to
+  // change the state to running and unblock the DB instances.
+  while (!ShouldStall() && !queue_.empty()) {
+    std::unique_lock<std::mutex> lock(mu_);
+    DB* db = queue_.front();
+    queue_.pop_front();
+    (static_cast<DBImpl*>(db))->GetWBMStallInterface()->SignalDB();
+  }
+}
+
+bool WriteBufferManager::ShouldStall() {
+  if (enabled()) {
```

Comment:
I don't think we want to impact existing behavior. Its better if the stall behavior is enabled through an option in the ```WriteBufferManager``` constructor. And if its enabled, I think the stall condition should only be ```memory_usage() >= buffer_size_```, since the idea is to have a hard limit. Also, once the stall is triggered, it might be a good idea to have a low watermark for removing the stall in order to give some time for flushes to complete.